ipld/kamt/src/pointer.rs

```diff
@@ -79,66 +71,51 @@ where
     where
         S: Serializer,
     {
+        use serde::ser::SerializeMap;
+
         // Using a `Map` for everything to keep things simple.
         // Constructing the map manually so we don't have to clone the extension and give it to a struct.
-        let mut map = BTreeMap::new();
+        let mut map = serializer.serialize_map(Some(1))?;
         match self {
             Pointer::Values(vals) => {
-                add_to_ipld_map::<S, _>(&mut map, TAG_VALUES, vals)?;
+                map.serialize_entry("v", vals)?;
             }
             Pointer::Link { cid, ext, .. } => {
-                add_to_ipld_map::<S, _>(&mut map, TAG_LINK, &(cid, ext))?;
+                map.serialize_entry("l", &(cid, ext.len(), BytesSer(ext.path_bytes())))?;
             }
             Pointer::Dirty { .. } => {
                 return Err(ser::Error::custom("Cannot serialize cached values"))
             }
         }
-        Ipld::Map(map).serialize(serializer)
-    }
-}
-
-impl<K, V, H, const N: usize> TryFrom<Ipld> for Pointer<K, V, H, N>
-where
-    K: DeserializeOwned,
-    V: DeserializeOwned,
-{
-    type Error = String;
-
-    fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
-        if let Ipld::Map(mut map) = ipld {
-            if let Some(values) = from_ipld_map::<Vec<KeyValuePair<K, V>>>(&mut map, TAG_VALUES) {
-                return values.map(Self::Values);
-            }
-
-            if let Some(link) = from_ipld_map::<(Cid, Extension)>(&mut map, TAG_LINK) {
-                return link.map(|(cid, ext)| Self::Link {
-                    cid,
-                    ext,
-                    cache: Default::default(),
-                });
-            }
-
-            Err(format!(
-                "Expected pointer tag in map; got {:#?}",
-                map.keys()
-            ))
-        } else {
-            Err(format!("Expected `Ipld::Map`, got {:#?}", ipld))
-        }
+        map.end()
     }
 }
 
 /// Deserialize the Pointer like a tagged enum.
 impl<'de, K, V, H, const N: usize> Deserialize<'de> for Pointer<K, V, H, N>
 where
-    K: DeserializeOwned,
-    V: DeserializeOwned,
+    K: Deserialize<'de>,
+    V: Deserialize<'de>,
 {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        Ipld::deserialize(deserializer).and_then(|ipld| ipld.try_into().map_err(de::Error::custom))
+        #[derive(Deserialize)]
+        enum PointerDe<K, V> {
+            #[serde(rename = "v")]
+            Values(Vec<KeyValuePair<K, V>>),
+            #[serde(rename = "l")]
+            Link(Cid, u32, BytesDe),
+        }
+        match PointerDe::<K, V>::deserialize(deserializer)? {
+            PointerDe::Values(vals) => Ok(Self::Values(vals)),
+            PointerDe::Link(k, l, BytesDe(p)) => Ok(Self::Link {
+                cid: k,
+                ext: Extension::new(l, p),
+                cache: Default::default(),
+            }),
+        }
```

Comment:
To be honest I liked the way it was better because it was symmetrical: we manually inserted things into a map and then we manually parsed it out using the tags and the types, approximating what I'm guessing the derive macro generates. Now we manually fill the map using the serializer API but use derivation to pull it back out. It begs the question: why not use these DTO structs for serialization as well? I didn't want to use such technical structs to avoid having to allocate them even as temporary structures, although arguably tuples must be allocated too. 

Isn't there a way to use the serde API to look for the tags? Maybe it's just my sense of beauty, but it seems a bit arbitrary this way.

Reply:
> why not use these DTO structs for serialization as well?

We can, it's just more code. Unfortunately, I couldn't use the _same_ struct due to borrowing/cloning/copying.

> Isn't there a way to use the serde API to look for the tags? Maybe it's just my sense of beauty, but it seems a bit arbitrary this way.

What kind of tags? CBOR? NOTE: the current logic _isn't_ CBOR specific, I'm just avoiding constructing temporary maps.

Reply:
I've changed this code to be the same in both cases, avoiding the manual logic.

Reply:
>  I didn't want to use such technical structs to avoid having to allocate them even as temporary structures, although arguably tuples must be allocated too.

This will involve a _tiny_ bit of memory copying, but not much. Honestly, it should just get optimized away and inlined.

Unfortunately, the _serialize_ side is easy but the _deserialize_ side is more painful to implement manually.

But note: using _maps_ was definitely slower because that actually has to allocate.

Reply:
By tags I meant `"v"` and `"l"`, and whatever the opposite of `serialize_map` is.

Now that you have changed it to have two versions of the DTOs it's easier to see why it can't be done with just one :+1: